source/as-promise.js

```diff
@@ -27,28 +28,34 @@ module.exports = options => {
 				return;
 			}
 
-			const {statusCode} = response;
 			const limitStatusCode = options.followRedirect ? 299 : 399;
 
 			response.body = data;
 
 			try {
 				for (const hook of options.hooks.afterResponse) {
 					// eslint-disable-next-line no-await-in-loop
-					response = await hook(response);
-
-					if (is.plainObject(response)) {
-						if (emitter.retry(response) === false) {
-							reject(new Error('Retry limit reached.'));
+					response = await hook(response, updatedOptions => {
+						if (emitter.retry(null) === false) {
+							throw new Error('Retry limit reached.');
 						}
```

Comment:
I don't fully understand this code, but this request should not be affected by the retry limit.

Reply:
Then it'd be possible to have infinite loop. Assume it always returns 401, no matter what. Then it keeps retrying, and retrying... Keep in mind that the hooks are still present.

Reply:
So this only let it run the hook once?

Reply:
No, the hook is ran anyway, no matter if is it a normal request or a retry. But if it fails anyway, it loops. And to prevent the infinite loop we need to check the retries.

Reply:
https://github.com/sindresorhus/got/pull/647/commits/98e51810797c8c93cc0b2fe55f5d0e6d2e7eb687#diff-0920de5c123084fabef750aa0967f8d4R309

Reply:
I think this behavior needs to be documented.

Reply:
Or we can stop executing next hooks in the retry, then we won't need `emitter.retry(null) == false`. Example:

hook 1
hook 2 - retries
	hook 1
hook 3

Current behavior (very bad, it shouldn't be done like that):
hook 1
hook 2 - retries
	hook 1
	hook 2
	hook 3
hook3

Reply:
> Alright. I think this behavior needs to be documented though.

No need to have that behavior. Fixed! There's no infinity loop. See my previous comment :)